lib/spack/spack/environment.py

```diff
@@ -1057,6 +1232,48 @@ def __exit__(self, exc_type, exc_val, exc_tb):
             activate(self._previous_active)
 
 
+def _concretize_from_constraints(spec_constraints):
+    # Accept only valid constraints from list and concretize spec
+    # Get the named spec even if out of order
+    root_spec = [s for s in spec_constraints if s.name]
+    if len(root_spec) != 1:
+        m = 'Spec %s is not a valid concretization target' % s.name
```

Comment:
This is hitting a python 2 vs. python 3 difference, as in python 3 variables defined in a list comprehension will not leak into the enclosing scope. I found this reference in [footnote 1](https://docs.python.org/2/reference/expressions.html#id20) that says:

> In Python 2.3 and later releases, a list comprehension “leaks” the control variables of each for it contains into the containing scope. However, this behavior is deprecated, and relying on it will not work in Python 3